src/sentry/models/outbox.py

```diff
@@ -0,0 +1,378 @@
+        # we should simply accept the occasional multiple sends than to introduce hard locking.
+        # so long as all objects sent are committed, and so long as any concurrent changes to data
+        # result in a future processing, we should always converge on non stale values.
+        coalesced: OutboxBase | None = self.select_coalesced_objects().last()
+        yield coalesced
+        if coalesced is not None:
+            self.select_coalesced_objects().filter(id__lte=coalesced.id).delete()
+
+    def process(self) -> bool:
+        with self.process_coalesced() as coalesced:
+            if coalesced is not None:
+                coalesced.send_signal()
+                return True
+        return False
+
+    @abc.abstractmethod
+    def send_signal(self):
+        pass
+
+    def drain_shard(self, max_updates_to_drain: int | None = None):
+        runs = 0
+        while max_updates_to_drain is None or runs < max_updates_to_drain:
+            runs += 1
+            next_row: OutboxBase | None = self.select_sharded_objects().first()
+            if next_row is None:
+                return True
+            next_row.process()
+
+        if self.select_sharded_objects().first() is not None:
+            raise OutboxFlushError(
+                f"Could not flush items from shard {self.key_from(self.sharding_columns)!r}"
+            )
+
+
+MONOLITH_REGION_NAME = "--monolith--"
+
+
+# Outboxes bound from region silo -> control silo
+@region_silo_only_model
+class RegionOutbox(OutboxBase):
+    def send_signal(self):
+        process_region_outbox.send(
+            sender=OutboxCategory(self.category),
+            payload=self.payload,
+            object_identifier=self.object_identifier,
+        )
+
+    sharding_columns = ("scope", "scope_identifier")
+    coalesced_columns = ("scope", "scope_identifier", "category", "object_identifier")
+
+    class Meta:
+        app_label = "sentry"
+        db_table = "sentry_regionoutbox"
+        index_together = (
+            (
+                "scope",
+                "scope_identifier",
+                "category",
+                "object_identifier",
+            ),
+            (
+                "scope",
+                "scope_identifier",
+                "scheduled_for",
+            ),
+            ("scope", "scope_identifier", "id"),
+        )
+
+    __repr__ = sane_repr("scope", "scope_identifier", "category", "object_identifier")
+
+    @classmethod
+    def drain_for_model(cls, model_inst: Any, max_updates_to_drain: int | None = None) -> None:
+        outbox = cls.for_model_update(model_inst)
+        if outbox is None:
+            return
+        outbox.drain_shard(max_updates_to_drain)
+
+    @classmethod
+    def for_model_update(cls, model_inst: Any) -> RegionOutbox:
+        from sentry.models import Organization, OrganizationMember
+
+        result = cls()
+        if isinstance(model_inst, Organization):
```

Comment:
I could move this logic into the individual model classes, yes.

Autofilling isn't possible -- the scope, scope identifier are highly dependent on semantic of how we intend to replicate the object.  It isn't inferrable, it's a decision one carefully makes because it impacts the sharding and ordering of replication.

object_identifier could be autofilled.  Category, might?  But it's just a mapping between model and value, basically.  And the "known values" need to be provided ahead of time, explicitly.  I also do not prefer it to be a string or some namespace that maps to model names directly.